### PR TITLE
Check executor params coherence

### LINKED ISSUE #1704: Stricter checks for `ExecutorParams` soundness
As rightly [noted](https://github.com/paritytech/polkadot-sdk/pull/1192#discussion_r1335599938) by @mrcnski, people make mistakes, and the governance consists of people. Although the `ExecutorParams` restrictions are properly documented, a situation may arise when they are not totally sound. Some probable concerns are:
* Inclusion of multiple occurrences of the same `ExecutorParam` variant, in which case only the first one will take effect;
* Inclusion of a parameter value that overcomes limits imposed by the underlying code (e.g., `PrecheckingMaxMemory` is `u64`, but if its value is over `isize::MAX`, a panic will occur);
* Inclusion of senseless values (e.g., a backing timeout longer than the block time).

`ExecutorParams` are currently set as a part of `HostConfiguration`, and there's already a mechanism to check for the configuration soundness that allows refusing an unsound configuration: [the consistency check](https://github.com/paritytech/polkadot-sdk/blob/f209b31b48586bca81c0ffafa150e1796ee05845/polkadot/runtime/parachains/src/configuration.rs#L362). It should be complemented with appropriate checks for `ExecutorParams` soundness.

### PR DESCRIPTION
**Pull request by eagr**
closes #1704

Really appreciate the help and patience from @s0me0ne-unkn0wn 🙏
I left some unsolved questions in the FIXME comments, would you take a look?

kusama address: FvpsvV1GQAAbwqX6oyRjemgdKV11QU5bXsMg9xsonD1FLGK

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3866703

**Review by s0me0ne-unkn0wn — CHANGES_REQUESTED** [review]
Thanks for the contribution, it's highly appreciated! Left some comments on how to improve it.

@eskimor @sandreim @slumber we need some brainstorming here. We want to limit values of backing and approval timeouts that could be set via `ExecutorParams` to some reasonable range, that is, to forbid setting a value that breaks things. Technically, for backing timeout, we could limit from 0 to `block_time`, and that's good enough to rule out obvious errors in configuration. Practically, I'd like to have more reasonable values, as setting the backing limit to 0, or, say, 0.01 sec, definitely breaks everything. My first guess is to limit it from 1/12 blocktime to 1/2 blocktime but I may not have a solid enough understanding of underlying logic, especially after async backing has landed. Do you have better ideas?

File: polkadot/primitives/src/v6/executor_params.rs
Comment (eagr) [reply polkadot/primitives/src/v6/executor_params.rs:264]: Yeah, I guess this could the policy for the rest of params that if we don't have a good candidate for a bound.

File: polkadot/primitives/src/v6/executor_params.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/primitives/src/v6/executor_params.rs:32]: Valid point. Okay, we can do it the other way. We can stop accounting for those additional pages in the limit (set the limit to (0, 65536] as it was in the beginning). Instead, we can always add those 32 pages to `*max_pages` [here](https://github.com/paritytech/polkadot-sdk/blob/aad80cce318fd492ce0d9c9951f2d7ec949710f3/polkadot/node/core/pvf/common/src/executor_intf.rs#L110C59-L110C59) (with `.saturating_add()` for the sake of correctness). Thus, the limit set on-chain becomes pure Wasm linear memory limit, implicitly increased by those additional pages. In that case, no constant moving is needed.

File: polkadot/primitives/src/v6/executor_params.rs
Comment (eagr) [reply polkadot/primitives/src/v6/executor_params.rs:174]: I did think about using HashMap coz I also need to cache the values, but couldn't seem to import it easlily. Thanks for the heads up! I'll try BTreeMap.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3871745

File: polkadot/primitives/src/v6/executor_params.rs
Comment (eagr) [reply polkadot/primitives/src/v6/executor_params.rs]: Using literals until then.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3873605

File: polkadot/primitives/src/v6/executor_params.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/primitives/src/v6/executor_params.rs]: For the next couple of weeks, people are mostly OOO, so I don't expect quick responses... Let's just make it, I don't see how someone would object to this change; the timeouts are Polkadot's internal kitchen, and it's quite logical for them to be placed with the primitives.

File: polkadot/primitives/src/v6/mod.rs
Comment (s0me0ne-unkn0wn) [inline polkadot/primitives/src/v6/mod.rs:1571]: Looks like your formatter is adding a bunch of unsolicited corrections... It won't pass formatting check in CI like this. Are you using the latest nightly toolchain and `cargo +nightly fmt`?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3887439

**Review by s0me0ne-unkn0wn — COMMENTED** [review]
Looks good! Now waiting for other teammates to come back and discuss the exact values of limits.

**Review by s0me0ne-unkn0wn — APPROVED** [review]

**Comment by eagr** [comment]
In the mean time, any good next issue I could work on?

**Comment by s0me0ne-unkn0wn** [comment]
Hmm, maybe #677 (somewhat connected with this one), but we'd still need @mrcnski there.

Also, there's a test improvement puzzle #1597. It would be good to have it solved, but it's not urgent and nobody has time to even start thinking about it.

File: polkadot/primitives/src/v6/executor_params.rs
Comment (mrcnski) [inline polkadot/primitives/src/v6/executor_params.rs:187]: Can we have this logic right in `configuration.rs`? This file seems to be mainly defining a bunch of types with not much actual logic.

File: polkadot/primitives/src/v6/executor_params.rs
Comment (mrcnski) [inline polkadot/primitives/src/v6/executor_params.rs:207]: Maybe should be renamed to `OutsideLimit`.

File: polkadot/primitives/src/v6/executor_params.rs
Comment (mrcnski) [inline polkadot/primitives/src/v6/executor_params.rs:71]: Can we reference the constant names instead of having these numbers in the documentation? There's a high risk of the docs and constants going out of sync.

File: polkadot/primitives/src/v6/executor_params.rs
Comment (mrcnski) [inline polkadot/primitives/src/v6/executor_params.rs:193]: I'm loving the checks you added, but is there a reason we use a macro instead of a simple fn here?

File: polkadot/primitives/src/v6/executor_params.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/primitives/src/v6/executor_params.rs:43]: Ah yes, it was split into `PvfPrepTimeoutKind` and `PvfExecTimeoutKind` long ago, but docs wasn't updated

```suggestion
// See also `PvfPrepTimeoutKind` and `PvfExecTimeoutKind` docs.
```

File: polkadot/node/core/pvf/common/src/executor_intf.rs
Comment (mrcnski) [inline polkadot/node/core/pvf/common/src/executor_intf.rs:113]: Nice! But can we just use the value from `sem` instead?

**Review by mrcnski — CHANGES_REQUESTED** [review]
Great stuff, much appreciate your time!

File: polkadot/primitives/src/v6/executor_params.rs
Comment (eagr) [reply polkadot/primitives/src/v6/executor_params.rs:187]: Yeah, it would be nice to if we could put the checks in one place. But that may require opening up the internal of `ExecutorParams`, do you think it's worth it? @mrcnski

File: polkadot/primitives/src/v6/executor_params.rs
Comment (eagr) [reply polkadot/primitives/src/v6/executor_params.rs:193]: The function implementation might be more complicated: more inputs, branches, and more jumps. Rough idea:

```
fn check_param(seen, name, val, predicate) -> Result
check(seen, name, val, predicate)?
```

Using macro here seems simpler.

File: polkadot/node/core/pvf/common/src/executor_intf.rs
Comment (eagr) [reply polkadot/node/core/pvf/common/src/executor_intf.rs:49]: I see it's being used by the executo worker. You mean?

File: polkadot/node/core/pvf/common/src/executor_intf.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/node/core/pvf/common/src/executor_intf.rs:113]: I think we discussed it [here](https://github.com/paritytech/polkadot-sdk/pull/1774#discussion_r1343770041)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3928083

File: polkadot/node/core/pvf/common/src/executor_intf.rs
Comment (mrcnski) [reply polkadot/node/core/pvf/common/src/executor_intf.rs:49]: I mean why have `DEFAULT_NATIVE_STACK_MAX` and `NATIVE_STACK_MAX` when they are exactly the same? 🤔

File: polkadot/primitives/src/v6/executor_params.rs
Comment (mrcnski) [reply polkadot/primitives/src/v6/executor_params.rs:201]: Not sure what you mean exactly, but I *think* it's fine.

File: polkadot/node/core/pvf/common/src/executor_intf.rs
Comment (mrcnski) [reply polkadot/node/core/pvf/common/src/executor_intf.rs:113]: Oh I misread this code, nevermind. Should it be this instead though?

```suggestion
					maximum_pages: Some((*max_pages).saturating_add(EXTRA_HEAP_PAGES)),
```

And are these values correct? I would have thought the extra pages were 32. @s0me0ne-unkn0wn 

```rs
const DEFAULT_HEAP_PAGES_ESTIMATE: u32 = 32;
const EXTRA_HEAP_PAGES: u32 = 2048;
```

Comment (s0me0ne-unkn0wn) [reply polkadot/node/core/pvf/common/src/executor_intf.rs:113]: No, I believe everything is okay with the current version.

`EXTRA_HEAP_PAGES` is the number of pages of linear memory the instance is allowed to allocate in addition to its initial memory size.

`DEFAULT_HEAP_PAGES_ESTIMATE` is the number of pages the instance will use for its own purposes, they are not part of linear memory.

So, the corresponding executor parameter defines the number of linear memory pages the instance is allowed to allocate, and here, we implicitly add those auxiliary pages to that limit (for example, we can set the limit to 0, and Wasm code not using any linear memory should still be functional, but it's not possible without having some space for shadow stack and other bits).

**Review by s0me0ne-unkn0wn — APPROVED** [review]

File: polkadot/primitives/src/v6/executor_params.rs
Comment (eagr) [reply polkadot/primitives/src/v6/executor_params.rs:201]: Sorry, I was thinking that we might need to make the internal state public in order to do the checks in another crate, like `pub struct ExecutorParams(pub Vec<ExecutorParam>)`. Then it's possible that other crates may rely on the internal state which is not great.

I just realized that it comes with an iterator :)

But still, for the sake of separation of concerns, I think it wouldn't be a bad idea that the type does its own checks. Like, what if the checks are needed else where (other than `configurations.rs`)?

File: polkadot/node/core/pvf/common/src/executor_intf.rs
Comment (eagr) [reply polkadot/node/core/pvf/common/src/executor_intf.rs:113]: Would it clarify stuff if we name it `SHADOW_STACK_HEAP_PAGES`? Then current name may be a bit vague.

File: polkadot/node/core/pvf/common/src/executor_intf.rs
Comment (eagr) [reply polkadot/node/core/pvf/common/src/executor_intf.rs:49]: aye aye :)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3942649

File: polkadot/primitives/src/v6/executor_params.rs
Comment (mrcnski) [reply polkadot/primitives/src/v6/executor_params.rs:201]: Fair enough. :)

**Review by mrcnski — APPROVED** [review]
Thanks! ⭐️

File: polkadot/node/core/pvf/common/src/executor_intf.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/node/core/pvf/common/src/executor_intf.rs:110]: It's not just shadow stack, according to the comment :thinking: But I doubt if the reasoning behind the calculation provided is correct enough. I didn't dig deeper into this aspect of Wasmtime, but it seems obvious that data segment pages shouldn't be allocated implicitly. Wasm source must specify the initial memory size, and it should always be enough to accommodate the data segment at least. Considering that I'm not the author of the original constant naming, I'm not 100% sure what is better, but probably renaming `DEFAULT_HEAP_PAGES_ESTIMATE` to `INITIAL_HEAP_PAGES_ESTIMATE` would bring more clarity.

**Comment by eagr** [comment]
Thank you guys for spending the time to review! Is this good to go now?

**Pull request merged**
